cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -112,63 +126,116 @@ func createKubeConfigFiles(outDir string, cfg *kubeadmapi.InitConfiguration, kub
 	return nil
 }
 
-// getKubeConfigSpecs returns all KubeConfigSpecs actualized to the context of the current InitConfiguration
-// NB. this methods holds the information about how kubeadm creates kubeconfig files.
-func getKubeConfigSpecs(cfg *kubeadmapi.InitConfiguration) (map[string]*kubeConfigSpec, error) {
+// createKubeConfigAndCSRFiles creates all the requested kubeconfig files.
+// If kubeconfig files already exists, they are used only if evaluated equal; otherwise an error is returned.
+// This is a variant for use in CreateJoinControlPlaneKubeConfigAndCSRFiles which generates
+// CSR files instead of signed certificates.
+func createKubeConfigAndCSRFiles(kubeConfigDir string, cfg *kubeadmapi.InitConfiguration, kubeConfigFileNames ...string) error {
 
-	caCert, caKey, err := pkiutil.TryLoadCertAndKeyFromDisk(cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName)
+	// gets the KubeConfigSpecs, actualized for the current InitConfiguration
+	specs, err := getKubeConfigSpecsBase(cfg)
 	if err != nil {
-		return nil, errors.Wrap(err, "couldn't create a kubeconfig; the CA files couldn't be loaded")
+		return err
 	}
 
+	for _, kubeConfigFileName := range kubeConfigFileNames {
+		// retrieves the KubeConfigSpec for given kubeConfigFileName
+		spec, exists := specs[kubeConfigFileName]
+		if !exists {
+			return errors.Errorf("couldn't retrieve KubeConfigSpec for %s", kubeConfigFileName)
+		}
+
+		// builds the KubeConfig object
+		config, csr, err := buildKubeConfigAndCSRFromSpec(spec, cfg.ClusterName)
+		if err != nil {
+			return err
+		}
+
+		// writes the kubeconfig to disk if it not exists
+		if err = createKubeConfigFileIfNotExists(kubeConfigDir, kubeConfigFileName, config); err != nil {
+			return err
+		}
+		// Write CSR to disk
+		if err := pkiutil.WriteCSR(kubeConfigDir, kubeConfigFileName, csr); err != nil {
+			return errors.Wrap(err, "failed to write CSR file")
```

Comment:
returning the error here without Wrap is fine as the underlying errors are descriptive:
https://github.com/kubernetes/kubernetes/blob/4e79344501bfc5d3bc48caa12c41b9246d90850d/cmd/kubeadm/app/util/pkiutil/pki_helpers.go#L166-L181